### ISSUE TITLE
Problem of test data features extraction

### ISSUE THREAD

**Issue by csitaula** [jlmelville/uwot#12]
I reduced the dimension using supervised method. and  I used metric learning to reduce the dimension of the test data. But, the accuracy of training feature is 99%, whereas the testing data has very low accuracy, why? Any idea in this case. I don't want to use PCA or other tools for dimension reduction. Thanks
Below is the code!
library(uwot)
library(tidyverse)
library(ggplot2)
# iris_umap <- umap(iris, n_neighbors = 40, alpha = 0.6, init = "random",,n_threads =12,n_components = 
#perform dimension reduction for object detection features
data_train <- read.table('d://augmented_features/raw_training_features.csv',sep = ',',header = F)
data_test <- read.table('d://augmented_features/raw_testing_features.csv',sep = ',',header = F)
umap_test <- data_test[,-ncol(data_test)]
umap_train <- data_train[,-ncol(data_train)]
train_label <- data_train$V513
#set.seed(1337)
reduced_umap_train <- umap(umap_train,ret_model = TRUE,y=train_label,n_components = 2)
inria_train <- as.data.frame(reduced_umap_train$embedding)
inria_train %>% 
  mutate(Categories = data_train$V513) %>%
  ggplot(aes(V1, V2, color = Categories)) + geom_point(cex=1.5)
#write this file for final feature extraction
#class(inria_umap)
#write.tainble(inria_umap,'G://testing.csv',sep = ',',row.names = F,col.names = F)
set.seed(1337)
reduced_umap_test <- umap_transform(umap_test,reduced_umap_train)
inria_test <- as.data.frame(reduced_umap_test)
inria_test %>% 
  mutate(Categories = data_test$V513) %>%
  ggplot(aes(V1, V2, color = Categories)) + geom_point(cex=1.5)

#make final reduced datasets that is used for SVM training
category <- data_train$V513
train_data <- cbind(inria_train,category)
category <- data_test$V513
test_data <- cbind(inria_test,category)

#write data into the file
write.table(train_data,'d://augmented_features/reduced_training.csv',sep = ',',row.names = F,col.names = F)
write.table(test_data,'d:/augmented_features/reduced_testing.csv',sep = ',',row.names = F,col.names = F)

**Comment by jlmelville** [comment]
You seem to be using `umap` and `umap_transform` in the correct way, so there's no error there. My guess would be that the supervised UMAP is overfitting. Presumably you see good results when you plot `inria_train`? How does the `inria_test` plot compare?

Two things to try:

* Increase the dimension of the space you embed into by increasing `n_components`.
* Make UMAP less supervised by reducing `target_weight`. It takes a value between `0` (only uses the data in `X`, so just like UMAP without specifying `y`) and `1` (only uses the data in `y`), and the default is `0.5`, so set it to a value less than `0.5` and see what happens. It may be that supervised UMAP isn't helpful if you're going to pass the results into another supervised machine learning method.

**Comment by csitaula** [comment]
Yes, I found good accuracy in SVM for inria_train for two class problem (99%) and without using labels, the accuracy is 83%, which is again acceptable for this type of dataset which I am getting from other methods as well. But, for inria_test, the accuracy is around 82% using metric learning. I got this accuracy in SVM and the clusters are also odd type for inria_test while plotting, too. While not using the supervised method, it is okay! The accuracy lies in between 80 to 84% in total. I need to use a supervised approach. I will try your suggestions! Lets see!
